### PR TITLE
periodically check whether assigned kubelet config should become last-known-good

### LINKED ISSUE #57808: Fix Kubelet last-known-good config graduation
The last-known-good is only updated on Kubelet restart.

We could have this scenario:

lkg is set to local
we set config A
config A passes trial period, but nothing caused Kubelet to restart
we set config B, which turns out to be invalid
Kubelet will fall back to local, because lkg was never updated

### PR DESCRIPTION
**Pull request by mtaufen**
Fixes #57808

Previously, the last-known-good was only updated on Kubelet restart. This has been on my todo list for a while, good to finally have a PR up.

Previously we could have this scenario, which is fixed by this PR:
- lkg is set to local
- we set config A
- config A passes trial period, but nothing caused Kubelet to restart
-  we set config B, which turns out to be invalid
- Kubelet will fall back to local, because lkg was never updated

```release-note
NONE
```

### CODE REVIEW

**Comment by mtaufen** [comment]
removed hold

File: pkg/kubelet/kubeletconfig/controller.go [hunk 1]
```diff
@@ -142,15 +142,17 @@ func (cc *Controller) Bootstrap() (*kubeletconfig.KubeletConfiguration, error) {
...
-		} else if !trial {
-			utillog.Infof("assigned config passed trial period, will set as last-known-good")
-			if err := cc.graduateAssignedToLastKnownGood(); err != nil {
-				utillog.Errorf("failed to set last-known-good to assigned config, error: %v", err)
+		// update the last-known-good config if necessary, and start a timer that
+		// periodically checks whether the last-known good needs to be updated
+		// we only do this when the assigned config loads and passes validation
+		duration := assigned.ConfigTrialDuration.Duration
+		cc.checkTrial(duration)
+		go func() {
```
Comment (Random-Liu) [inline pkg/kubelet/kubeletconfig/controller.go]: nit: Use `go wait.Until(function, period, wait.NeverStop)`?

**Comment by Random-Liu** [comment]
LGTM with one nit.

File: pkg/kubelet/kubeletconfig/controller.go (hunk 1 above)
Comment (mtaufen) [reply pkg/kubelet/kubeletconfig/controller.go]: good idea

**Comment by mtaufen** [comment]
for some reason have a kubeadm file in the diff, let me see...

**Comment by mtaufen** [comment]
Ah, need to rebase on the 2018 fix

**Comment by mtaufen** [comment]
Fixed

**Comment by mtaufen** [comment]
re-applying label after removing extraneous generated files

**Pull request merged**